File: go/profiles.go
package basehuman

const (
	frozenBodyAlphabet     = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
	frozenChecksumAlphabet = "234679ACDEFGHJKMNPQRTUVWXY"
)

func frozenAliases() map[string]string {
	return map[string]string{"O": "0", "I": "1", "L": "1"}
}

// Baseh32V1Profile returns the frozen assisted-support profile baseh32-v1:
// 6 body plus 1 checksum symbol with the feistel-v1 permutation and no
// separators. Key material is application-specific and never part of the
// frozen profile; see spec 7.4.
func Baseh32V1Profile(keyBytes []byte, keyID string) Profile {
	return Profile{
		ProfileID:        "baseh32-v1",
		BodyAlphabet:     frozenBodyAlphabet,
		BodyLength:       6,
		ChecksumAlphabet: frozenChecksumAlphabet,
		ChecksumLength:   1,
		CaseSensitive:    false,
		Separator:        "",
		Grouping:         nil,
		Aliases:          frozenAliases(),
		Permutation: Permutation{
			Enabled:   true,
			Algorithm: "feistel-v1",
			KeyID:     keyID,
			KeyBytes:  keyBytes,
			Rounds:    8,
		},
	}
}

// Baseh32SV1Profile returns the frozen self-service profile baseh32s-v1:
// 6 body plus 2 checksum symbols. Detects all single-symbol substitutions
// and all adjacent transpositions; see spec 6.3.
func Baseh32SV1Profile(keyBytes []byte, keyID string) Profile {
	p := Baseh32V1Profile(keyBytes, keyID)
	p.ProfileID = "baseh32s-v1"
	p.ChecksumLength = 2
	return p
}

File: rust/src/profiles.rs
//! Frozen profiles (spec section 17).

use crate::profile::{Permutation, Profile};

const BODY_ALPHABET: &str = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
const CHECKSUM_ALPHABET: &str = "234679ACDEFGHJKMNPQRTUVWXY";
const DEFAULT_ROUNDS: u32 = 8;

const ALIASES: &[(char, char)] = &[('O', '0'), ('I', '1'), ('L', '1')];

fn frozen(profile_id: &str, checksum_length: usize, key_bytes: &[u8], key_id: &str) -> Profile {
    Profile {
        profile_id: profile_id.to_string(),
        body_alphabet: BODY_ALPHABET.to_string(),
        body_length: 6,
        checksum_alphabet: CHECKSUM_ALPHABET.to_string(),
        checksum_length,
        case_sensitive: false,
        separator: String::new(),
        grouping: Vec::new(),
        aliases: ALIASES.to_vec(),
        permutation: Permutation::FeistelV1 {
            key_id: key_id.to_string(),
            key_bytes: key_bytes.to_vec(),
            rounds: DEFAULT_ROUNDS,
        },
        profanity: None,
    }
}

/// Frozen profile `baseh32-v1`: 6 body + 1 checksum, feistel-v1 permutation,
/// 8 rounds. Assisted-support use; structured single-substitution miss rate
/// about 1.2 percent per position (spec 6.3).
///
/// Applications assign their own `key_id` and key material; both must be
/// immutable for the life of the profile.
pub fn baseh32_v1(key_bytes: &[u8], key_id: &str) -> Profile {
    frozen("baseh32-v1", 1, key_bytes, key_id)
}

/// Frozen profile `baseh32s-v1`: 6 body + 2 checksum, feistel-v1
/// permutation, 8 rounds. Self-service use; provably detects all
/// single-symbol substitutions and all adjacent transpositions (spec 6.3).
pub fn baseh32s_v1(key_bytes: &[u8], key_id: &str) -> Profile {
    frozen("baseh32s-v1", 2, key_bytes, key_id)
}
